overlord/snapstate/snapstate.go

```diff
@@ -652,6 +652,15 @@ func TryPath(st *state.State, name, path string, flags Flags) (*state.TaskSet, e
 //
 // The returned TaskSet will contain a DownloadAndChecksDoneEdge.
 func Install(st *state.State, name, channel string, revision snap.Revision, userID int, flags Flags) (*state.TaskSet, error) {
+	return InstallUnderDeviceContext(st, name, channel, revision, userID, flags, nil)
+}
+
+// InstallUnderDeviceContext returns a set of tasks for installing snap.
+// It will query for the snap under the given deviceCtx.
+// Note that the state must be locked by the caller.
+//
+// The returned TaskSet will contain a DownloadAndChecksDoneEdge.
+func InstallUnderDeviceContext(st *state.State, name, channel string, revision snap.Revision, userID int, flags Flags, deviceCtx DeviceContext) (*state.TaskSet, error) {
```

Comment:
Also, it's taking 7 parameters now. Maybe we could think about refactoring this a bit once the PRs land. Perhaps something like:
```
type Request struct {
    Name string
    Channel string
    Revision snap.Revision
}

type RequestContext {
    DeviceCtx DeviceContext
    UserID int
    Flags Flags
}

func Install(st *State, req Request, reqCtx RequestContext) 
func Update(st *State, req Request, reqCtx RequestContext)
```

Reply:
Yes, John has a similar plan, at least for the first part of the signature, that's also why I'll probably leave it alone for now.